test/core/async_job/queue_test.exs

```diff
@@ -159,35 +164,81 @@ defmodule AntikytheraCore.AsyncJob.QueueTest do
-    job_id = "foobar"
-    assert Queue.cancel(@queue_name, job_id) == {:error, :not_found}
+  test "cancel nonexisting job" do
+    assert Queue.cancel(@queue_name, @job_id) == {:error, :not_found}
+  end
 
-    # cancel waiting job
+  test "cancel waiting job" do
     second_after = Time.now() |> Time.shift_seconds(1)
-    assert register_job([id: job_id, schedule: {:once, second_after}]) == :ok
-    {:ok, status1} = Queue.status(@queue_name, job_id)
-    assert status1.state                     == :waiting
-    assert Queue.cancel(@queue_name, job_id) == :ok
-    assert Queue.cancel(@queue_name, job_id) == {:error, :not_found}
+    assert register_job([id: @job_id, schedule: {:once, second_after}]) == :ok
+    {:ok, status1} = Queue.status(@queue_name, @job_id)
+    assert status1.state                      == :waiting
+    assert Queue.cancel(@queue_name, @job_id) == :ok
+    assert Queue.cancel(@queue_name, @job_id) == {:error, :not_found}
+  end
 
-    # cancel runnable job
+  test "cancel runnable job" do
     assert Queue.fetch_job(@queue_name) == nil
-    assert register_job([id: job_id])  == :ok
+    assert register_job([id: @job_id])  == :ok
     assert Queue.start_jobs_and_get_metrics(Process.whereis(@queue_name))
     assert_receive({:"$gen_cast", :job_registered})
-    {:ok, status2} = Queue.status(@queue_name, job_id)
-    assert status2.state                     == :runnable
-    assert Queue.cancel(@queue_name, job_id) == :ok
-    assert Queue.cancel(@queue_name, job_id) == {:error, :not_found}
+    {:ok, status2} = Queue.status(@queue_name, @job_id)
+    assert status2.state                      == :runnable
+    assert Queue.cancel(@queue_name, @job_id) == :ok
+    assert Queue.cancel(@queue_name, @job_id) == {:error, :not_found}
+  end
 
-    # cancel running job
-    assert register_job([id: job_id])  == :ok
+  test "cancel running job" do
+    assert register_job([id: @job_id]) == :ok
     {_key, _job} = Queue.fetch_job(@queue_name)
-    {:ok, status3} = Queue.status(@queue_name, job_id)
-    assert status3.state                     == :running
-    assert Queue.cancel(@queue_name, job_id) == :ok
-    assert Queue.cancel(@queue_name, job_id) == {:error, :not_found}
+    {:ok, status3} = Queue.status(@queue_name, @job_id)
+    assert status3.state                      == :running
+    assert Queue.cancel(@queue_name, @job_id) == :ok
+    assert Queue.cancel(@queue_name, @job_id) == {:error, :not_found}
+  end
+
+  test "consecutive registrations should be rejected by rate limit" do
+    n = div(RateLimit.max_tokens(), RateLimit.tokens_per_command())
+    Enum.each(1..n, fn _ ->
+      assert register_job([]) == :ok
+    end)
+    {:error, {:rate_limit_reached, _}} = register_job([])
+  end
+
+  test "consecutive cancels should be rejected by rate limit" do
+    n = div(RateLimit.max_tokens(), RateLimit.tokens_per_command())
+    Enum.each(0..n, fn i ->
+      AsyncJobHelper.reset_rate_limit_status(@epool_id)
```

Comment:
Any intent behind calling `AsyncJobHelper.reset_rate_limit_status/1` on every iteration?
Since it looks to me that we can do this:
```ex
Enum.each(1..n, fn i ->
  assert register_job([job_id: @job_id <> "#{i}"]) == :ok
end)
AsyncJobHelper.reset_rate_limit_status(@epool_id)
assert register_job([job_id: @job_id <> "0"]) == :ok
Enum.each(1..n, fn i ->
  assert Queue.cancel(@queue_name, @job_id <> "#{i}")
end)
{:error, {:rate_limit_reached, _}} = Queue.cancel(@queue_name, @job_id <> "0")
```

Reply:
I've fixed some of my silly mistakes (missing `== :ok` and wrong key name `:job_id` => `:id`).
(Thank you for making me aware of them)
I'm not particularly concerned about how we "reset" the token bucket; as you showed less function call is better.
And actually we can make the tests a bit more compact by sharing the test setup for job cancellation and assertions for job registration.
https://github.com/access-company/antikythera/pull/8/commits/48b65c2543bf3d37737e79d94b3edd1c62d6898e